src/runtime/manifest-primitive-parser/manifest-primitive-parser.ts

```diff
@@ -0,0 +1,150 @@
+/**
+ * @license
+ * Copyright (c) 2017 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {Loader} from '../../platform/loader.js';
+import {parse as parser} from '../../gen/runtime/manifest-parser.js';
+
+import * as AstNode from '../manifest-ast-types/manifest-ast-nodes.js';
+import {Dictionary} from '../../utils/lib-utils.js';
+
+export type Ast = AstNode.All[];
+export {AstNode};
+
+interface ManifestLoadOptions {
+  registry?: Dictionary<Promise<Ast>>;
+  errors?: ManifestError[];
+}
+
+export interface ManifestParseOptions extends ManifestLoadOptions {
+  filename?: string;
+  loader?: Loader;
+}
+
+export enum ErrorSeverity {
+  Error = 'error',
+  Warning = 'warning'
+}
+
+export class ManifestError extends Error {
+  key: string;
+  location: AstNode.SourceLocation;
+  severity = ErrorSeverity.Error;
+  constructor(location: AstNode.SourceLocation, message: string) {
+    super(message);
+    this.location = location;
+  }
+}
+
+export class ManifestWarning extends ManifestError {
+  severity = ErrorSeverity.Warning;
+}
+
+export class ManifestPrimitiveParser {
+  static async load(path: string, loader: Loader, options: ManifestLoadOptions = {}): Promise<Ast> {
+    if (!loader) {
+      throw new Error('loader is required to load manifest for parsing');
+    }
+    // TODO(sjmiles):
+    // (1) IMO caching should be done at the Loader level
+    // (2) unless 'path' is normalized, it's not necessarily a key (could be more than one path to the same file)
+    const {registry} = options;
+    if (registry && registry[path]) {
+      return await registry[path];
+    }
+    const errors = options.errors || [];
+    const content = await loader.loadResource(path);
+    const promise = this.parse(content, {filename: path, loader, registry, errors});
+    if (registry) {
+      registry[path] = promise;
+    }
+    return await promise;
+  }
+  static async parse(content: string, options: ManifestParseOptions = {}): Promise<Ast> {
+    const {filename} = options;
+    let items: Ast = [];
+    try {
+      items = parser(content, {filename}) as Ast;
+    } catch (e) {
+      throw this.processError(new ManifestError(e.location, e.message), content, options);
+    }
+    await this.parseImports(items, options);
+    return items;
+  }
+  static extract(kind: string, fromAst: Ast) {
```

Comment:
This bit is tricky because it's extracting types dynamically. My typescript-fu is not strong. 

If I produce `Ast`, it will not cast to `RecipeNode[]`, for example, requiring this:
` const recipe = ManifestPrimitiveParser.extract('recipe', manifest)[0] as unknown as RecipeNode;`

I can produce `unknown[]` and then it's castable. 
` const recipe = ManifestPrimitiveParser.extract('recipe', manifest)[0] as RecipeNode;`

Alternatively, we could create extractors (by hand) for all possible AST types, but I didn't want to go there at least until I examined the deeper `Manifest` api.

Other ideas?

Reply:
Actually I think it's just a mistake that `RecipeNode` (and `Schema`) aren't included in `AstNode.All`. Recipes and schemas are both definitely AST nodes, so I'd suggest changing that instead and making this return `AstNode.All[]`.

We absolutely don't want extractors per AST type.